test/39-compose-images.spec.ts

```diff
@@ -0,0 +1,530 @@
+			},
+		];
+		const IMAGE_TO_REMOVE = {
+			id: 246,
+			name:
+				'registry2.balena-cloud.com/v2/793f9296017bbfe026334820ab56bb3a@sha256:2c969a1ba1c6bc10df53481f48c6a74dbd562cfb41ba58f81beabd03facf5582',
+			appId: 1658654,
+			serviceId: 650325,
+			serviceName: 'app_1',
+			imageId: 2693229,
+			releaseId: 1524186,
+			dependent: 0,
+		};
+		const images = {
+			'registry2.balena-cloud.com/v2/793f9296017bbfe026334820ab56bb3a@sha256:2c969a1ba1c6bc10df53481f48c6a74dbd562cfb41ba58f81beabd03facf5582': {
+				Containers: -1,
+				Created: 1599492712,
+				Id:
+					'sha256:f1154d76c731f04711e5856b6e6858730e3023d9113124900ac65c2ccc90e8e7',
+				Labels: {
+					'io.balena.architecture': 'aarch64',
+					'io.balena.device-type': 'jetson-tx2',
+					'io.balena.qemu.version': '4.0.0+balena2-aarch64',
+				},
+				ParentId: '',
+				RepoDigests: [
+					'registry2.balena-cloud.com/v2/793f9296017bbfe026334820ab56bb3a@sha256:72b80cbd3cc12de08d4adc9dec79916bf466031553f55b59c29e397829ea129f',
+				],
+				RepoTags: [
+					'registry2.balena-cloud.com/v2/793f9296017bbfe026334820ab56bb3a:delta-ada9fbb57d90e61e',
+				],
+				SharedSize: -1,
+				Size: 217024648,
+				VirtualSize: 217024648,
+			},
+			'livepush-supervisor:11.12.11': {
+				Containers: -1,
+				Created: 1599492599,
+				Id:
+					'sha256:db5af2c94366275d8e6d7ea3047f2405eab2f04a27f66843634a45958ef59f5a',
+				Labels: {
+					'io.balena.architecture': 'aarch64',
+					'io.balena.livepush-image': '1',
+					'io.balena.qemu.version': '4.0.0+balena-aarch64',
+				},
+				ParentId:
+					'sha256:4acb1652178e72c0b6143a08225e0df5ef74b338a0c9e2ca4cd261339f4f0431',
+				RepoDigests: null,
+				RepoTags: ['livepush-supervisor:11.12.11'],
+				SharedSize: -1,
+				Size: 501352885,
+				VirtualSize: 501352885,
+			},
+		};
+		// Stub the database to return images we want
+		mockedDatabase.setImages(IMAGES_FROM_DB).stub();
+		// Perform the test with our specially crafted data
+		await mockedDockerode.testWithData({ images }, async () => {
+			// Check that our legacy image exists
+			await expect(docker.getImage(IMAGE_TO_REMOVE.name).inspect()).to
+				.eventually.not.be.undefined;
+			await expect(
+				db.models('image').select().where(IMAGE_TO_REMOVE),
+			).to.eventually.have.lengthOf(1);
+			// Check that docker has this Image
+			await expect(docker.getImage(IMAGE_TO_REMOVE.name).inspect()).to
+				.eventually.not.be.undefined;
+			// Now remove this image...
+			await Images.remove(IMAGE_TO_REMOVE);
+			// Check if it still exists!
+			await expect(docker.getImage(IMAGE_TO_REMOVE.name).inspect()).to
+				.eventually.be.undefined;
+			await expect(db.models('image').select().where(IMAGE_TO_REMOVE)).to
+				.eventually.be.empty;
+			// Check that docker remove was called once
+			const removeSteps = _(mockedDockerode.actions)
+				.pickBy({ name: 'remove' })
+				.map()
+				.value();
+			expect(removeSteps).to.have.lengthOf(1);
+		});
+	});
+
+	it('Removes a single Image', async () => {
```

Comment:
This uses real data that causes this bug to actually happen. We are adding this test coverage precisely because it caused us to write a new PR to fix that bug and this is the exact payloads that will cause it.